### PR TITLE
`ConnectedElements` - duplicate elements bugfix, `Connectors` method added

### LINKED ISSUE #1476: Connected Elements - duplicate elements bug
#### Description:

`ConnectedElements` method returns duplicated elements.

#### Expected behaviour:

Method should return unique elements.

### PR DESCRIPTION
**Pull request by michal-pekacki**
### Issues addressed by this PR

Closes #1476 

Fixed duplicated elements bug.

### CODE REVIEW

File: Revit_Core_Engine/Query/ConnectedElements.cs [hunk 1]
```diff
@@ -45,24 +45,10 @@ public static List<Element> ConnectedElements(this Element element)
...
-                connSet = (element as MEPCurve).ConnectorManager?.Connectors;
-            }
-            
-            if (connSet == null)
-            {
-                BH.Engine.Base.Compute.RecordError("Input element is not supported by the method. Check if element is a valid MEP object.");
-                return null;
-            }
+            List<Connector> connectors = element.Connectors();
+            List<Element> connectedElements = new List<Element>();
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/ConnectedElements.cs:49]: You can make the method look more neat by using `HashSet<ElementId>`

```suggestion
            HashSet<ElementId> connectedElements = new HashSet<ElementId>();
```

...then:
```C#
(...)
 if (el != null && el.Id != element.Id)
    connectedElements.Add(el.Id);
```

File: Revit_Core_Engine/Query/ConnectedElements.cs [hunk 2]
```diff
@@ -73,7 +59,10 @@ public static List<Element> ConnectedElements(this Element element)
                 }
             }
 
-            return connectedElements.ToList();
+            //remove duplicates and host element
+            connectedElements = connectedElements.Where(x => x.Id.IntegerValue != element.Id.IntegerValue).GroupBy(x => x.Id.IntegerValue).Select(x => x.First()).ToList();
+
+            return connectedElements;
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/ConnectedElements.cs:65]:
```suggestion
            return connectedElements.ToList();
```

File: Revit_Core_Engine/Query/Connectors.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+
+        [Description("Returns connectors of the MEP element. MEPCurve connectors are sorted by distance from the startpoint, FamilyInstace connectors by IsPrimary property.")]
+        [Input("element", "MEP element to get the connectors from.")]
+        [Output("cconnectors", "MEP Element connectors.")]
+        public static List<Connector> Connectors(this Element element)
+        {
+            return SortedConnectors(element as dynamic);
+        }
+
+        /***************************************************/
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Connectors.cs:45]: Private methods header is missing 😉

File: Revit_Core_Engine/Query/Connectors.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("Returns connectors of the MEP element. MEPCurve connectors are sorted by distance from the startpoint, FamilyInstace connectors by IsPrimary property.")]
+        [Input("element", "MEP element to get the connectors from.")]
+        [Output("cconnectors", "MEP Element connectors.")]
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Connectors.cs:39]:
```suggestion
        [Output("connectors", "MEP Element connectors.")]
```

File: Revit_Core_Engine/Query/Connectors.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+        /***************************************************/
+
+        private static List<Connector> SortedConnectors(this MEPCurve mepCurve)
+        {
+            XYZ startPoint = (mepCurve.Location as LocationCurve).Curve.GetEndPoint(0);
+            ConnectorSet connSet = mepCurve.ConnectorManager?.Connectors;
+
+            List<Connector> connList = new List<Connector>();
+            foreach (Connector conn in connSet)
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Connectors.cs:53]: What if `connSet == null`?

File: Revit_Core_Engine/Query/Connectors.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        }
+
+        /***************************************************/
+
+        private static List<Connector> SortedConnectors(this FamilyInstance familyInstance)
+        {
+            ConnectorSet connSet = familyInstance.MEPModel?.ConnectorManager?.Connectors;
+
+            List<Connector> connList = new List<Connector>();
+            foreach (Connector conn in connSet)
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Connectors.cs:68]: What if `connSet == null`?

File: Revit_Core_Engine/Query/Connectors.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+                connList.Add(conn);
+
+            return connList;
+        }
+
+        /***************************************************/
+
+        private static List<Connector> SortedConnectors(this Element element)
+        {
+            BH.Engine.Base.Compute.RecordError("Input element is not supported by the method. Check if element is a valid MEP object.");
```
Comment (pawelbaran) [inline Revit_Core_Engine/Query/Connectors.cs:78]:
```suggestion
            BH.Engine.Base.Compute.RecordError("Input element is not supported by the connector query method. Check if element is a valid MEP object.");
```

**Review by pawelbaran — CHANGES_REQUESTED** [review]

File: Revit_Core_Engine/Query/Connectors.cs (hunk 5 above)
Comment (michal-pekacki) [reply Revit_Core_Engine/Query/Connectors.cs:53]: Null check added 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check compliance

**Review by pawelbaran — APPROVED** [review]
Happy to approve, changes make sense and work based on the review session with @michal-pekacki 👍

**Comment by michal-pekacki** [comment]
@BHoMBot check required

**Comment by michal-pekacki** [comment]
@BHoMBot check ready-to-merge

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**